File: apps/agent-macos/Sources/EgressViewAgentCore/ThirdPartyGeoLookup.swift
import Foundation

/// Asks a third party where an address is, for the addresses nothing else can
/// name.
///
/// **This is the only path that sends a watched destination out of the
/// network**, which is why it runs only when someone has chosen it. The
/// settings screen and the README have said so since before it existed --
/// until 2026-09-13 the toggle was wired to nothing at all, so a reader who
/// turned it on got no lookups and no warning that none were happening
/// (P3-115).
///
/// The service is `ipwho.is`. The first attempt used `ip-api.com`, which the
/// README had named for months: its free tier answers over plain HTTP only,
/// App Transport Security refuses that, and every lookup failed with -1022
/// before it left the Mac. Measured 2026-09-13: `https://ip-api.com/` returns
/// 403 (TLS is a paid feature), `https://ipwho.is/` returns the location.
/// Sending watched addresses in clear text was not a trade worth making for a
/// tool whose purpose is showing what leaves the machine.
public struct ThirdPartyGeoLookup: Sendable {
    /// The free tier has no bulk endpoint -- that is a paid feature -- so this
    /// is one request per address, and the count matters. Free allows 1,000
    /// requests a day; half of that is the most this will spend, so a day of
    /// unusual browsing cannot exhaust the allowance and leave the map blank.
    public static let dailyBudget = 500
    /// Addresses per run. Enough for a burst of new destinations, small enough
    /// that one run cannot spend the day's budget.
    public static let batchSize = 25
    /// One request a second. The published limits are far looser; a monitoring
    /// tool that gets itself rate-limited stops answering the question it
    /// exists to answer.
    public static let minimumInterval: TimeInterval = 1

    public struct Located: Equatable, Sendable {
        public let ip: String
        public let countryCode: String
        public let latitude: Double
        public let longitude: Double
        public let city: String?
    }

    public protocol Transport: Sendable {
        func get(_ request: URLRequest) async throws -> (Data, HTTPURLResponse)
    }

    public enum Failure: Error, Equatable {
        case httpStatus(Int)
        case malformedResponse
    }

    private let base: URL
    private let transport: any Transport
    private let sleep: @Sendable (TimeInterval) async -> Void

    public init(
        base: URL = URL(string: "https://ipwho.is")!,
        transport: any Transport,
        sleep: @escaping @Sendable (TimeInterval) async -> Void = { seconds in
            try? await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
        }
    ) {
        self.base = base
        self.transport = transport
        self.sleep = sleep
    }

    /// The address that is asked about, and nothing else.
    func request(for address: String) -> URLRequest? {
        guard var components = URLComponents(
            url: base.appendingPathComponent(address), resolvingAgainstBaseURL: false
        ) else { return nil }
        components.queryItems = [
            URLQueryItem(name: "fields", value: "success,ip,country_code,latitude,longitude,city"),
        ]
        guard let url = components.url else { return nil }
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        return request
    }

    /// Looks up addresses one at a time and returns only the ones located.
    ///
    /// An address the service cannot place is left out rather than stored as
    /// somewhere: a wrong country on the map is worse than a missing one,
    /// because nothing tells the reader it is wrong.
    ///
    /// A failure ends the run rather than marching through the rest. The free
    /// tier carries no availability guarantee, so the service being down is an
    /// ordinary Tuesday, and the remaining addresses are still in the queue
    /// for the next run.
    public func locate(_ addresses: [String], budget: Int = batchSize) async throws -> [Located] {
        let wanted = Array(addresses.prefix(max(0, min(budget, Self.batchSize))))
        guard !wanted.isEmpty else { return [] }
        var located: [Located] = []
        for (index, address) in wanted.enumerated() {
            guard let request = request(for: address) else { continue }
            if index > 0 { await sleep(Self.minimumInterval) }
            let (data, response) = try await transport.get(request)
            guard response.statusCode == 200 else { throw Failure.httpStatus(response.statusCode) }
            guard let row = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                throw Failure.malformedResponse
            }
            if let place = Self.located(row, requested: address) { located.append(place) }
        }
        return located
    }

    static func located(_ row: [String: Any], requested: String) -> Located? {
        guard (row["success"] as? Bool) == true,
              let country = row["country_code"] as? String, country.count == 2,
              let latitude = row["latitude"] as? Double,
              let longitude = row["longitude"] as? Double
        else { return nil }
        // The reply echoes the address, but the row is stored against the
        // address that was asked about either way.
        let ip = (row["ip"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? requested
        let city = (row["city"] as? String).flatMap { $0.isEmpty ? nil : $0 }
        return Located(
            ip: ip, countryCode: country.uppercased(),
            latitude: latitude, longitude: longitude, city: city
        )
    }
}

public struct URLSessionThirdPartyGeoTransport: ThirdPartyGeoLookup.Transport {
    private let session: URLSession

    public init(session: URLSession = .shared) { self.session = session }

    public func get(_ request: URLRequest) async throws -> (Data, HTTPURLResponse) {
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw ThirdPartyGeoLookup.Failure.malformedResponse
        }
        return (data, http)
    }
}

File: apps/agent-windows/src/EgressView.Agent.Core/ThirdPartyGeoLookup.cs
using System.Text.Json;

namespace EgressView.Agent.Core;

/// Where to look when the location cache does not have an address.
public enum GeoLookupSource
{
    /// Only what the last fetch brought. Nothing is asked for.
    CacheOnly,

    /// Ask the Hub again, off-schedule. It has usually resolved the address
    /// already, and asking it sends nothing outside the network the Hub is on.
    Hub,

    /// The Hub first, then a third party for what the Hub does not know.
    ///
    /// <b>This is the only path that sends a watched address out of the
    /// network.</b> It stays off unless someone chooses it.
    HubThenThirdParty,
}

public static class GeoLookupSources
{
    public static string ToWire(this GeoLookupSource source) => source switch
    {
        GeoLookupSource.CacheOnly => "cache-only",
        GeoLookupSource.Hub => "hub",
        _ => "hub-then-third-party",
    };

    public static GeoLookupSource Parse(string? wire) => wire switch
    {
        "cache-only" => GeoLookupSource.CacheOnly,
        "hub-then-third-party" => GeoLookupSource.HubThenThirdParty,
        _ => GeoLookupSource.Hub,
    };

    public static bool UsesHub(this GeoLookupSource source) => source != GeoLookupSource.CacheOnly;
    public static bool UsesThirdParty(this GeoLookupSource source) => source == GeoLookupSource.HubThenThirdParty;
}

/// Asks a third party where an address is, for the addresses nothing else can
/// name.
///
/// <b>This is the only path that sends a watched destination out of the
/// network</b>, which is why it runs only when someone has chosen it.
///
/// The service is `ipwho.is`, over HTTPS. The Mac's first attempt used
/// `ip-api.com`, whose free tier answers over plain HTTP only; sending watched
/// addresses in clear text was not a trade worth making for a tool whose
/// purpose is showing what leaves the machine.
public sealed class ThirdPartyGeoLookup(HttpClient? http = null, Uri? baseUri = null, Func<TimeSpan, CancellationToken, Task>? delay = null)
{
    /// The free tier has no bulk endpoint, so this is one request per address
    /// and the count matters. Free allows 1,000 requests a day; half of that is
    /// the most this will spend, so a day of unusual browsing cannot exhaust
    /// the allowance and leave the map blank.
    public const int DailyBudget = 500;

    /// Addresses per run. Enough for a burst of new destinations, small enough
    /// that one run cannot spend the day's budget.
    public const int BatchSize = 25;

    /// One request a second. The published limits are far looser; a monitoring
    /// tool that gets itself rate-limited stops answering the question it
    /// exists to answer.
    public static readonly TimeSpan MinimumInterval = TimeSpan.FromSeconds(1);

    private readonly HttpClient http = http ?? CreateClient();
    private readonly Uri baseUri = baseUri ?? new Uri("https://ipwho.is");
    private readonly Func<TimeSpan, CancellationToken, Task> delay = delay ?? Task.Delay;

    private static HttpClient CreateClient()
    {
        var client = new HttpClient { Timeout = TimeSpan.FromSeconds(20) };
        AgentEnrollmentClient.EnsureUserAgent(client);
        return client;
    }

    /// The address that is asked about, and nothing else.
    public Uri RequestUri(string address) =>
        new($"{baseUri.AbsoluteUri.TrimEnd('/')}/{Uri.EscapeDataString(address)}" +
            "?fields=success,ip,country_code,latitude,longitude,city");

    /// Looks addresses up one at a time and returns only the ones located.
    ///
    /// An address the service cannot place is left out rather than stored as
    /// somewhere: a wrong country on the map is worse than a missing one,
    /// because nothing tells the reader it is wrong.
    ///
    /// A failure ends the run rather than marching through the rest. The free
    /// tier carries no availability guarantee, so the service being down is an
    /// ordinary Tuesday, and the remaining addresses are still there next time.
    public async Task<IReadOnlyList<GeoLocation>> LookUpAsync(
        IReadOnlyList<string> addresses, int budget, CancellationToken cancellationToken = default)
    {
        var located = new List<GeoLocation>();
        var spent = 0;
        foreach (var address in addresses)
        {
            if (spent >= budget || spent >= BatchSize) break;
            if (spent > 0) await delay(MinimumInterval, cancellationToken);
            spent++;
            GeoLocation? location;
            try { location = await LookUpOneAsync(address, cancellationToken); }
            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested) { throw; }
            catch (Exception) { break; }
            if (location is not null) located.Add(location);
        }
        Spent = spent;
        return located;
    }

    /// How many requests the last run actually made, including the ones that
    /// found nothing. The budget is spent by asking, not by being answered.
    public int Spent { get; private set; }

    private async Task<GeoLocation?> LookUpOneAsync(string address, CancellationToken token)
    {
        using var response = await http.GetAsync(RequestUri(address), token);
        if (!response.IsSuccessStatusCode) throw new HttpRequestException($"HTTP {(int)response.StatusCode}");
        using var document = JsonDocument.Parse(await response.Content.ReadAsByteArrayAsync(token));
        var root = document.RootElement;
        if (root.ValueKind != JsonValueKind.Object) return null;
        if (root.TryGetProperty("success", out var success) && success.ValueKind == JsonValueKind.False) return null;
        if (!root.TryGetProperty("latitude", out var lat) || !root.TryGetProperty("longitude", out var lon)) return null;
        if (!lat.TryGetDouble(out var latitude) || !lon.TryGetDouble(out var longitude)) return null;
        var country = root.TryGetProperty("country_code", out var code) && code.ValueKind == JsonValueKind.String
            ? code.GetString() : null;
        var city = root.TryGetProperty("city", out var name) && name.ValueKind == JsonValueKind.String
            ? name.GetString() : null;
        // The reply names the address it answered about. Trusting the reply's
        // own idea of which address this was would let a confused answer write
        // a location against a destination nobody asked about.
        return new GeoLocation(address, latitude, longitude, country, city);
    }
}
